db/migrate/20180410092453_policy_log.rb

```diff
@@ -0,0 +1,87 @@
+Sequel.migration do
+  tables = %i(roles role_memberships resources permissions annotations)
+
+  up do
+    execute """
+      CREATE TYPE policy_log_op AS ENUM ('INSERT', 'DELETE', 'UPDATE');
+      CREATE TYPE policy_log_kind AS ENUM #{literal tables.map(&:to_s)};
+      CREATE EXTENSION IF NOT EXISTS hstore;
+    """
+
+    key = %i(policy_id version)
+    create_table :policy_log do
+      String :policy_id, null: false
+      Integer :version, null: false
+      foreign_key key, :policy_versions, on_delete: :cascade
+      index key
+
+      column :operation, :policy_log_op, null: false
+      column :kind, :policy_log_kind, null: false
+      column :subject, :hstore, null: false
+      column :at, :timestamptz, null: false, default: Sequel.function(:clock_timestamp)
+    end
+
+    tables.each do |table|
+      # find the primary key of the table
+      primary_key = schema(table).select{|x,s|s[:primary_key]}.map(&:first).map(&:to_s).pg_array
+      execute """
+        CREATE OR REPLACE FUNCTION policy_log_#{table}() RETURNS TRIGGER AS $$
+          DECLARE
+            subject #{table};
+            current policy_versions;
+          BEGIN
+            IF (TG_OP = 'DELETE') THEN
+              subject := OLD;
+            ELSE
+              subject := NEW;
+            END IF;
+            current = current_policy_version();
+            IF current.resource_id = subject.policy_id THEN
+              INSERT INTO policy_log(
+                policy_id, version,
+                operation, kind,
+                subject)
+              SELECT
+                current.resource_id, current.version,
+                TG_OP::policy_log_op, '#{table}'::policy_log_kind,
```

Comment:
Singularize `kind`.
`role_memberships` as a `kind` of operation sounds too SQL-y?
The `subject` of the insert includes all the insert data? That's not intuitive from the field name.
Subject data as `resource_id` seems SQL-y. Would prefer simply `resource`.
Role and/or privilege grant should include the `owner` field.
We need to know the role which initiated each action.
We need to be sure that we can index the roles and resources as in v4.
`operation` as all-caps is odd compared to the other fields.
We should take the opportunity to use cucumbers as the definitive specification of audit data formats.
We should compare the audit data formats to CyberArk audit, to incorporate the learnings and best practices.
IP address is not being captured.

What's the syntax `TG_OP::policy_log_op`?
How does a table know its `policy_log_kind`?

For my edification: What does slice on an hstore and `literal primary_key` do? What does `literal` do?

Reply:
This is meant to be SQL-y. It's just the log of DB operations. It's going to be pushed out to a separate storage and querying backend. Which role initiated each action is already stored in `policy_versions`.

`TG_OP::policy_log_op` is a cast to `policy_log_op` type. Table doesn't know its `policy_log_kind`, it's embedded in the text of the trigger function. `literal primary_key` transforms the `primary_key` to SQL literal, eg. it makes `%w(foo bar)` into `('foo', 'bar')`.